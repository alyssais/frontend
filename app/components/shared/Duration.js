// @flow weak

import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import 'intersection-observer';
import VisibilityObserver from 'react-intersection-observer';
import { second } from 'metrick/duration';
import moment from 'moment';
import { getDuration, getDurationString } from 'app/lib/date';

type DurationFormats = 'full' | 'medium' | 'short' | 'micro';

type Props = {
  from?: (moment$Moment | string | number | Date | number[]),
  to?: (moment$Moment | string | number | Date | number[]),
  className?: string,
  tabularNumerals: boolean,
  format: DurationFormats,
  updateFrequency?: number
};

type State = {
  seconds?: number
};

// Grab a copy of the Moment constructor so we can test PropTypes
const Moment = moment().constructor;

function makeDurationFormat(format: DurationFormats) {
  const Component = (props: Props) => <Duration {...props} format={format} />;
  Component.displayName = format.charAt(0).toUpperCase() + format.slice(1);
  return Component;
}

export default class Duration extends React.PureComponent<Props, State> {
  static Full = makeDurationFormat('full');
  static Medium = makeDurationFormat('medium');
  static Short = makeDurationFormat('short');
  static Micro = makeDurationFormat('micro');

  static propTypes = {
    from: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.instanceOf(Date),
      PropTypes.instanceOf(Moment)
    ]),
    to: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.instanceOf(Date),
      PropTypes.instanceOf(Moment)
    ]),
    className: PropTypes.string,
    tabularNumerals: PropTypes.bool.isRequired,
    format: PropTypes.oneOf(getDurationString.formats),
    updateFrequency: PropTypes.number
  };

  static defaultProps = {
    updateFrequency: second.bind(1),
    tabularNumerals: true
  };

  _timeout: TimeoutID;

  state = {};

  tick() {
    const { from, to } = this.props;
    const seconds = getDuration(from, to).asSeconds();

    this.setState({ seconds }, this.maybeScheduleTick);
  }

  componentDidMount() {
    this.maybeScheduleTick();
  }

  maybeScheduleTick() {
    const { from, to, updateFrequency } = this.props;

    // We only want to schedule ticks if our duration is indeterminate,
    // and our update frequency isn't zero
    if (!(from && to) && typeof updateFrequency == 'number' && updateFrequency > 0) {
      this._timeout = setTimeout(() => {
        this.tick();
      }, updateFrequency);
    }
  }

  maybeCancelTick() {
    if (this._timeout) {
      clearTimeout(this._timeout);
    }
  }

  componentWillUnmount() {
    this.maybeCancelTick();
  }

  static getDerivedStateFromProps(nextProps, currentState) {
    const seconds = getDuration(nextProps.from, nextProps.to).asSeconds();

    if (seconds !== currentState.seconds) {
      return { seconds };
    }

    return null;
  }

  componentDidUpdate(prevProps) {
    const { updateFrequency, to } = this.props;

    if (updateFrequency !== prevProps.updateFrequency || to !== prevProps.to) {
      this.maybeCancelTick();
      this.maybeScheduleTick();
    }
  }

  handleVisibilityChange = (visible) => {
    this.maybeCancelTick();

    if (visible) {
      this.tick();
    }
  }

  render() {
    const spanClassName = classNames(
      this.props.className,
      { 'tabular-numerals': this.props.tabularNumerals }
    );

    const durationString = getDurationString(this.state.seconds, this.props.format);

    if (this.props.to) {
      return (
        <span className={spanClassName}>
          {durationString}
        </span>
      );
    }

    return (
      <VisibilityObserver
        tag="span"
        className={spanClassName}
        onChange={this.handleVisibilityChange}
      >
        {durationString}
      </VisibilityObserver>
    );
  }
}
