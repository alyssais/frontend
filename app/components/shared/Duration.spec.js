/* global jest, describe, it, expect, xdescribe */
import React from 'react';
import ReactTestRenderer from 'react-test-renderer';

import Duration from './Duration';

const MOCKED_DURATION_SECONDS = 42069;
const MOCKED_DURATION = {
  asSeconds() {
    return MOCKED_DURATION_SECONDS;
  }
};

jest.mock('../../lib/date', () => {
  const getDurationString = jest.fn(() => 'MOCKED-DURATION');
  getDurationString.formats = ['expected'];
  const getDuration = jest.fn(() => MOCKED_DURATION);
  return { getDuration, getDurationString };
});

describe('Duration', () => {
  const componentList = Object.keys(Duration);
  const DurationComponent = Duration.Expected;

  it('provides stateless Components named after `getDurationString.formats`', () => {
    expect(componentList).toEqual(['Expected']);
  });

  it('calls through to `getDurationString`', () => {
    const getDuration = require('../../lib/date').getDuration;
    const getDurationString = require('../../lib/date').getDurationString;
    const from = new Date();
    const to = new Date();

    const component = ReactTestRenderer.create(
      <DurationComponent
        from={from}
        to={to}
        updateFrequency={0}
      />
    );

    const tree = component.toJSON();
    expect(getDuration).toHaveBeenCalledWith(from, to);
    expect(getDurationString).toHaveBeenCalledWith(MOCKED_DURATION_SECONDS, 'expected');
    expect(tree).toMatchSnapshot();
  });

  describe('tabularNumerals', () => {
    it('can be disabled', () => {
      const getDuration = require('../../lib/date').getDuration;
      const getDurationString = require('../../lib/date').getDurationString;
      const from = new Date();
      const to = new Date();

      const component = ReactTestRenderer.create(
        <DurationComponent
          tabularNumerals={false}
          from={from}
          to={to}
          updateFrequency={0}
        />
      );

      const tree = component.toJSON();
      expect(getDuration).toHaveBeenCalledWith(from, to);
      expect(getDurationString).toHaveBeenCalledWith(MOCKED_DURATION_SECONDS, 'expected');
      expect(tree).toMatchSnapshot();
    });
  });

  xdescribe('updateFrequency', () => {
    // TODO: Need to be able to instrument updating and interactions
    it('sets an interval if supplied with a frequency greater than zero', () => {
      const rendered = ReactTestRenderer.create(
        <DurationComponent
          from="2016-05-07T09:00:00.000Z"
          to="2016-05-07T09:00:00.000Z"
          updateFrequency={10}
        />
      );

      expect(rendered.toJSON()).toMatchSnapshot();
    });

    it('sets no interval if supplied with a frequency of zero', () => {
      const rendered = ReactTestRenderer.create(
        <DurationComponent
          from="2016-05-07T09:00:00.000Z"
          to="2016-05-07T09:00:00.000Z"
          updateFrequency={0}
        />
      );

      expect(rendered.toJSON()).toMatchSnapshot();
    });
  });
});
