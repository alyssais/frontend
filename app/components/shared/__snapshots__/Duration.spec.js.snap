// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Duration calls through to \`getDurationString\` 1`] = `
<span
  className="tabular-numerals"
>
  MOCKED-DURATION
</span>
`;

exports[`Duration tabularNumerals can be disabled 1`] = `
<span
  className=""
>
  MOCKED-DURATION
</span>
`;
